iidm/iidm-impl/src/main/java/com/powsybl/iidm/network/impl/AbstractConnectable.java

```diff
@@ -175,22 +175,40 @@ protected void move(TerminalExt oldTerminal, TopologyPoint oldTopologyPoint, Str
                 .build();
 
         // detach the terminal from its previous voltage level
-        attachTerminal(oldTerminal, oldTopologyPoint, (VoltageLevelExt) bus.getVoltageLevel(), terminalExt);
+        replaceTerminal(oldTerminal, ((VoltageLevelExt) bus.getVoltageLevel()).getTopologyModel(), terminalExt, true);
     }
 
-    private void attachTerminal(TerminalExt oldTerminal, TopologyPoint oldTopologyPoint, VoltageLevelExt voltageLevel, TerminalExt terminalExt) {
+    void replaceTerminal(TerminalExt oldTerminal, TopologyPoint oldTopologyPoint, TerminalExt newTerminalExt, boolean notify) {
+        Objects.requireNonNull(oldTerminal);
+        int iSide = terminals.indexOf(oldTerminal);
+        if (iSide == -1) {
+            throw new PowsyblException("Terminal to replace not found");
+        }
+        terminals.set(iSide, newTerminalExt);
+
+        if (newTerminalExt != null && notify) {
+            notifyUpdate("terminal" + (iSide + 1), oldTopologyPoint, newTerminalExt.getTopologyPoint());
+        }
+    }
+
+    void replaceTerminal(TerminalExt oldTerminal, TopologyModel newTopologyModel, TerminalExt newTerminalExt, boolean notify) {
         // first, attach new terminal to connectable and to voltage level of destination, to ensure that the new terminal is valid
-        terminalExt.setConnectable(this);
-        voltageLevel.getTopologyModel().attach(terminalExt, false);
+        if (newTerminalExt != null) {
```

Comment:
Couldn't we require newTerminalExt to be non-null? we don't need it anymore in this PR, and besides the referrer call for replacement is not working well in the implementations if newTerminalExt is null.

Reply:
Yes cleaned